File: src/main/java/alg01unionfind/UF4WeightedQuickUnionWithPathCompression.java
package alg01unionfind;

/**
 * <p>
 * Further improvement of union find. We compress paths while searching for roots of nodes.
 * <i>"Compressing"</i> means that we set all childs in a tree to tree's root.
 * </p>
 */
public class UF4WeightedQuickUnionWithPathCompression extends UF3WeightedQuickUnion implements UnionFindAlgorithm {

    public UF4WeightedQuickUnionWithPathCompression(int size) {
        super(size);
    }

    @Override
    protected int root(int i) {
        while (i != parentOf[i]) {
            // We bubble up node up to it's root to compress paths
            parentOf[i] = parentOf[parentOf[i]];
            // We cycle thought the node's parents until we find a node that is parent
            // to itself
            i = parentOf[i];
        }
        return i;
    }
}

File: src/main/kotlin/alg01unionfind/UF4WeightedQuickUnionWithPathCompression.kt
package alg01unionfind

/**
 * Another improvement for QuickUnion: compress tree paths by resetting node parents to tree roots
 * during root lookups.
 */
class UF4WeightedQuickUnionWithPathCompression(size: Int) : UF3WeightedQuickUnion(size) {
    override fun treeRoot(n: Int): Int {
        val root = super.treeRoot(n)
        var nWithNonRootParent = n
        // While the parent of a node in question is not actually root node it's remapped to root
        // to compress three paths.
        while (nWithNonRootParent != root) {
            val next_nWithNonRootParent = this.parentOf[nWithNonRootParent]
            this.parentOf[nWithNonRootParent] = root
            nWithNonRootParent = next_nWithNonRootParent
        }
        return root
    }
}